File: source/fortran/rebco_variables.f90
module rebco_variables
  !! author: M. Kovari
  !!
  !! Module for the REBCO HTS superconductor variables
  !!
  !! Variables relating to the REBCO HTS tape, strand and conductor
  !! Conduit information is in the modules relating to each coil.
  !!
  !!### References
  !!
  !! - Updated 13/11/18 using data from Lewandowska et al 2018.

#ifndef dp
  use, intrinsic :: iso_fortran_env, only: dp=>real64
#endif

  implicit none

  real(dp) :: rebco_thickness
  !! thickness of REBCO layer in tape (m) (`iteration variable 138`)

  real(dp) :: copper_thick
  !! thickness of copper layer in tape (m) (`iteration variable 139`)

  real(dp) :: hastelloy_thickness
  !! thickness of Hastelloy layer in tape (m)

  real(dp) :: tape_width
  !! Mean width of tape (m)

  real(dp) :: tape_thickness
  !! thickness of tape, inc. all layers (hts, copper, substrate, etc.) (m)

  real(dp) :: croco_od
  !! Outer diameter of CroCo strand (m)

  real(dp) :: croco_id
  !! Inner diameter of CroCo copper tube (m)

  real(dp) :: croco_thick
  !! Thickness of CroCo copper tube (m) (`iteration variable 158`)

  real(dp) :: copper_rrr
  !! residual resistivity ratio copper in TF superconducting cable

  real(dp) :: copperA_m2
  !! TF coil current / copper area (A/m2)

  real(dp) :: coppera_m2_max
  !! Maximum TF coil current / copper area (A/m2)

  real(dp) :: f_coppera_m2
  !! f-value for constraint 75: TF coil current / copper area < copperA_m2_max

  real(dp) :: copperaoh_m2
  !! CS coil current / copper area (A/m2) (`sweep variable 61`)

  real(dp) :: copperaoh_m2_max
  !! Maximum CS coil current / copper area (A/m2)

  real(dp) :: f_copperaoh_m2
  !! f-value for constraint 88: CS coil current / copper area < copperA_m2_max

  !#TODO: variables need descriptions and units
  real(dp) :: stack_thickness
  real(dp) :: tapes
  real(dp) :: rebco_area
  real(dp) :: copper_area
  real(dp) :: hastelloy_area
  real(dp) :: solder_area
  real(dp) :: croco_area
end module rebco_variables

File: process/data_structure/rebco_variables.py
from dataclasses import dataclass


@dataclass(slots=True)
class RebcoData:
    dx_tf_hts_tape_rebco: float = 1.0e-6
    """thickness of REBCO layer in tape (m) (`iteration variable 138`)"""

    dx_hts_tape_copper: float = 100.0e-6
    """thickness of copper layer in tape (m) (`iteration variable 139`)"""

    dx_hts_tape_hastelloy: float = 50.0e-6
    """thickness of Hastelloy layer in tape (m)"""

    dr_hts_tape: float = 4.0e-3
    """Mean width of tape (m)"""

    dx_hts_tape_total: float = 6.5e-5
    """thickness of tape, inc. all layers (hts, copper, substrate, etc.) (m)"""

    dia_croco_strand_tape_region: float = 0.0
    """Inner diameter of CroCo strand tape region (m)"""

    dx_croco_strand_copper: float = 2.5e-3
    """Thickness of CroCo strand copper tube (m) (`iteration variable 158`)"""

    copper_rrr: float = 100.0
    """residual resistivity ratio copper in TF superconducting cable"""

    coppera_m2: float = 0.0
    """TF coil current / copper area (A/m2)"""

    coppera_m2_max: float = 1.0e8
    """Maximum TF coil current / copper area (A/m2)"""

    copperaoh_m2: float = 0.0
    """CS coil current / copper area (A/m2) (`sweep variable 61`)"""

    copperaoh_m2_max: float = 1.0e8
    """Maximum CS coil current / copper area (A/m2)"""

    dx_croco_strand_tape_stack: float = 0.0
    """Width / thickness of tape stack in CroCo strand (m)"""

    n_croco_strand_hts_tapes: float = 0.0
    """Number of HTS tapes in CroCo strand"""

    a_croco_strand_rebco: float = 0.0
    """Area of REBCO in CroCo strand (m2)"""

    a_croco_strand_copper_total: float = 0.0
    """Area of copper in CroCo strand (includes tapes and outer tube) (m2)"""

    a_croco_strand_hastelloy: float = 0.0
    """Area of Hastelloy in CroCo strand (m2)"""

    a_croco_strand_solder: float = 0.0
    """Area of solder in CroCo strand (m2)"""

    a_croco_strand: float = 0.0
    """Total area of a CroCo strand (m2)"""


CREATE_DICTS_FROM_DATACLASS = RebcoData
